doc/config_yang_validation/config_db_yang_validation.md

```diff
@@ -0,0 +1,178 @@
+
+
+## Design
+
+### Design Summary
+Currently, updates to ConfigDB are published via 2 steps:
+1.	ad-hoc ConfigDB field validation 
+2.	Call to config_db.set_entry(target_value) which blindly pushes the target_value passed via connection to ConfigDB using ConfigDBConnector object
+
+Where YANG models are applicable, this project replaces these two steps with a ConfigDBConnector class decorator which validates the new config using preexisting YANG models before pushing the new config update to ConfigDB.
+
+Why use ConfigDBConnector class decorator:
+- 	allows us to preserve the preexisting instance of ConfigDBConnector (no new object initialization is required)
+- 	decorates only the necessary instances of ConfigDBConnector with added YANG validation
+- 	Non-invasive, all ConfigDBConnector code not writing to ConfigDB (no YANG validation necessary) can be left as-is
+
+### Decorator Graphical Depiction
+Existing ConfigDBConnector class and methods-
+
+![](../../images/config_yang_validation/decorator1.png)
+
+Decorator only overrides contents of original set_entry() method to include YANG validation
+
+![](../../images/config_yang_validation/decorator2.png)
+
+### YANG Validation Implementation
+The YANG validation included in our ConfigDBConnector set_entry() decorator leverages the preexisting SONiC Generic Config Updater library, which uses YANG to validate JSON patches that contain write requests to ConfigDB, and then pushes these updates to ConfigDB only after YANG validation passes. 
+
+Sample JSON patch:
+
+![](../../images/config_yang_validation/jsonpatch.png)
+
+### Decorator Pseudo Code
+```
+def validate_decorator(config_db_connector):
+        def validated_set_entry(updated_value):
+                update_json_patch = updated_value.to_json_patch()
+                GenericConfigUpdater.apply_patch(update_json_patch)
+        config_db_connector.set_entry = validated_set_entry
+
+```
+For the buffermgrd C++ use case, there are 3 options for YANG validation implementation:
+- Call into Python GCU code for YANG validation
+- Call into D-bus GCU APIs
+- C++ to invoke GCU command line
+## Additional Features
+
+### Validation Error Output - Exception Handling
+Currently, error messages for invalid write requests to ConfigDB are defined on a case-by-case basis. Our goal is to preserve the preexisting error messages, as Exceptions raised and default error messages output by Generic Config Updater YANG validation can be difficult for the client to comprehend. 
+
+To achieve this goal, we do the following:
+1.	Take note of the type of Exception raised for a particular error (e.g. GCU would raise a ValueError for an invalid ConfigDB field format)
+2.	From the calling function requesting a write to ConfigDB, catch the relevant exception type and output the original error message in exception handling
```

Comment:
How to handle if there are different error messages with same type of exception.

Reply:
I added a note to address this as well, thank you